### ISSUE TITLE
FancyArrowPatch picker fails depending on arrowstyle

### ISSUE THREAD

**Issue by lejar** [matplotlib/matplotlib#8384]
### Bug report

**Bug summary**

The picker of FancyArrowPatch will not work with some arrow styles (see included example). Certain arrow styles will cause the picker to produce a pick event if the user clicks anywhere to the left of the arrow.

**Code for reproduction**
```python
from matplotlib import pyplot as plt
from matplotlib import patches


def pick_event(event):
    print("PICK", event.artist)


def main():
    fig = plt.figure()
    axes = fig.add_subplot(111)
    axes.set_xlim((0, 10))
    axes.set_ylim((0, 10))

    arrow = patches.FancyArrowPatch(
        (8, 2),
        (9, 3),
        # Uncommenting this will make the pick event be triggered every time I
        # click to the left of the arrow.
        # arrowstyle='-|>',
    )
    # Same thing happens with arrow.set_picker(True)
    arrow.set_picker(4)

    fig.canvas.mpl_connect('pick_event', pick_event)
    axes.add_patch(arrow)
    plt.show()


if __name__ == '__main__':
    main()

```

**Actual outcome**
(With the arrowstyle line uncommented)
A pick event is always triggered when clicking to the left of the arrow.

```
PICK FancyArrowPatch(8,2->9,3)
PICK FancyArrowPatch(8,2->9,3)
PICK FancyArrowPatch(8,2->9,3)
PICK FancyArrowPatch(8,2->9,3)
PICK FancyArrowPatch(8,2->9,3)
```

**Expected outcome**
The pick event is not triggered unless the user actually clicks on the arrow.

**Matplotlib version**
I tested this with two versions of matplotlib:
- version 2.0.0 installed via pip 9.0.1 using Python 2.7.12 included in ubuntu
-  version 1.5.3 built from source using Python 2.7.11 built from source

OS information:
```bash
$ uname -a
Linux 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by phobson** [comment]
@lejar thanks for the detailed report and sorry for the slow response. hopefully one of us will be able to take a closer look at this soon.

**Comment by tacaswell** [comment]
Something is clearly wrong with the hit testing...

**Comment by tacaswell** [comment]
```python
from matplotlib import pyplot as plt
from matplotlib import patches

picks = []

def pick_event(event):
    picks.append(event)
    print("PICK", event.artist, event.mouseevent)



fig = plt.figure()
axes = fig.add_subplot(111)
axes.set_xlim((0, 10))
axes.set_ylim((0, 10))

arrow = patches.FancyArrowPatch(
    (8, 2),
    (9, 3),
    # Uncommenting this will make the pick event be triggered every time I
    # click to the left of the arrow.
    # arrowstyle='->',
)

arrow2 = patches.FancyArrowPatch(
    (8, 7),
    (9, 8),
    # Uncommenting this will make the pick event be triggered every time I
    # click to the left of the arrow.
    arrowstyle='->',
)
# Same thing happens with arrow.set_picker(True)
arrow.set_picker(4)
arrow2.set_picker(4)

fig.canvas.mpl_connect('pick_event', pick_event)
axes.add_patch(arrow)
axes.add_patch(arrow2)
plt.show()


def hit_mask(ax, paths):
    N, M = 200, 200
    X, Y = np.meshgrid(np.linspace(*ax.get_xlim(), N),
                       np.linspace(*ax.get_ylim(), M))
    res = np.zeros((N, M))
    for k, (x, y) in enumerate(zip(X.ravel(),
                                        Y.ravel())):
        for m, p in enumerate(paths):
            if p.contains_point((x, y), None, .1):
                res[np.unravel_index(k, (N, M))] += 2**m
    im = ax.imshow(res, extent=(X.min(), X.max(), Y.min(), Y.max()),
              origin='lower', cmap='Set3')
    im.get_cmap().set_under('w')
    im.set_clim(vmin=.5)

hit_mask(axes, [arrow.get_path(), arrow2.get_path()])
```
![so](https://cloud.githubusercontent.com/assets/199813/24504986/4e60b8d8-1526-11e7-9230-03af733c2ed1.png)

Sorry the code is messy, this is just what I have been playing with trying to debug this.

**Comment by QuLogic** [comment]
I think the second path is not closed?
```
(Pdb) p arrow.get_path()
Path(array([[ 8.03108293,  2.03445713],
       [ 8.49475998,  2.49813417],
       [ 8.95843702,  2.96181121],
       [ 8.95663002,  2.96506551],
       [ 8.96216088,  2.96637862],
       [ 8.96769173,  2.96769173],
       [ 8.96517254,  2.96095479],
       [ 8.96265335,  2.95421786],
       [ 8.96084635,  2.95747215],
       [ 8.49716931,  2.49379511],
       [ 8.03349226,  2.03011807],
       [ 8.03108293,  2.03445713],
       [ 8.03108293,  2.03445713]]), array([ 1,  3,  3,  2,  3,  3,  3,  3,  2,  3,  3,  2, 79], dtype=uint8))
```
vs.
```
(Pdb) p arrow2.get_path()
Path(array([[ 8.0322876 ,  7.0322876 ],
       [ 8.49998966,  7.49998966],
       [ 8.94961703,  7.94961703],
       [ 8.9407411 ,  7.94748949],
       [ 8.94961703,  7.94961703],
       [ 8.94555976,  7.93881137]]), array([1, 3, 3, 1, 2, 2], dtype=uint8))
```
Note the missing 79/CLOSEPOLY to close the path and the last point doesn't really seem to match the first.

**Comment by colonelsammy** [comment]
I have the same problem (Windows, built from source at 7d6b0c8).

Following your example:

```python
from matplotlib import pyplot as plt
from matplotlib import patches
import numpy as np

picks = []

def pick_event(event):
    picks.append(event)
    print("PICK", event.artist, event.mouseevent)

fig = plt.figure()
axes = fig.add_subplot(111)
axes.set_xlim((0, 10))
axes.set_ylim((0, 10))

arrow = patches.FancyArrowPatch(
    (8, 2),
    (9, 3),
    # Uncommenting this will make the pick event be triggered every time I
    # click to the left of the arrow.
    # arrowstyle='->',
)

arrow2 = patches.FancyArrowPatch(
    (8, 7),
    (9, 8),
    # Uncommenting this will make the pick event be triggered every time I
    # click to the left of the arrow.
    arrowstyle='->',
)
# Same thing happens with arrow.set_picker(True)
arrow.set_picker(4)
arrow2.set_picker(4)

p = arrow.get_path()
patch = patches.PathPatch(p, facecolor='orange', lw=2)
axes.add_patch(patch)

p = arrow2.get_path()
patch = patches.PathPatch(p, facecolor='orange', lw=2)
axes.add_patch(patch)

fig.canvas.mpl_connect('pick_event', pick_event)
axes.add_patch(arrow)
axes.add_patch(arrow2)
plt.show()

```
![image](https://cloud.githubusercontent.com/assets/703230/25916571/064589a4-35bd-11e7-8ded-af5800ac8f9a.png)

That second patch really doesn't look right!

**Comment by dstansby** [comment]
Perhaps the problem is that trying to pick an open patch (such as an open arrow doesn't make sense? (if a pick is defined as clicking inside the patch)

**Comment by cacti77** [comment]
Does anyone know when, or in which version of Matplotlib, this is likely to be fixed please? This issue is still in v2.2.2 on Windows with the arrowstyle values I really want to use, '->' or '-|>'.
My end-user doesn't like 'wedge', 'fancy' or 'simple'. If this issue isn't going to be fixed for a long time I might write my own picker function for use with '->' or '-|>'.

**Comment by anntzer** [comment]
AFAICT this was fixed by #10776.
